lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller/Gru.pm

```diff
@@ -0,0 +1,128 @@
+# (i.e. has task `obs_rsync_run` in 'inactive' state)
+#
+# QUEUE_FULL means that number of non-finished jobs for tasks 'obs_rsync_run' reached parameter `queue_limit`.
+# This condition shouldn't practically happen and means either overload or some bug in queueing.
+#
+# Code below does not have strong concurrency guards, but in worst case it should lead to project queued
+# twice, which shouldn't be a problem.
+use constant {
+    QUEUED     => 200,
+    STARTED    => 201,
+    IN_QUEUE   => 208,
+    QUEUE_FULL => 507,
+};
+
+my $lock_timeout = 36000;
+
+sub index {
+    my $self   = shift;
+    my $helper = $self->obs_rsync;
+    my %jobs;
+    my $results = $self->app->minion->backend->list_jobs(
+        0,
+        4 * $helper->queue_limit,
+        {tasks => ['obs_rsync_run'], states => ['active', 'inactive']});
+
+    for my $job (@{$results->{jobs}}) {
+        $jobs{$job->{id}} = _extend_job_info($job);
+    }
+
+    $self->render('ObsRsync_gru_index', jobs => \%jobs);
+}
+
+# this prepares fields for rendering
+sub _extend_job_info {
+    my ($job) = @_;
+
+    my $created_at = $job->{created};
+    $created_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($created_at)) if $created_at;
+
+    my $started_at = $job->{started};
+    $started_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($started_at)) if $started_at;
+
+    my $args = $job->{args};
+    if (ref $args eq 'HASH' && scalar(%$args) == 1 && $args->{project}) {
+        $args = $args->{project};
+    }
+    else {
+        $args = dump($args);
+    }
+    my $info = {
+        id       => $job->{id},
+        task     => $job->{task},
+        args     => $args,
+        created  => $created_at,
+        started  => $started_at,
+        priority => $job->{priority},
+        retries  => $job->{retries},
+        state    => $job->{state},
+        notes    => dump($job->{notes}),
+    };
+    return $info;
+}
+
+sub run {
+    my $self    = shift;
+    my $project = $self->param('folder');
+    my $helper  = $self->obs_rsync;
+    return undef if $helper->check_and_render_error($project);
+    my $app         = $self->app;
+    my $queue_limit = $helper->queue_limit;
+
+    my $results = $app->minion->backend->list_jobs(
+        0,
+        4 * $queue_limit,
+        {tasks => ['obs_rsync_run'], states => ['active', 'inactive']});
+
+    my $has_active_job = 0;
+    for my $job (@{$results->{jobs}}) {
+        if ($job->{args} && ($job->{args}[0]->{project} eq $project)) {
+            return $self->render(json => {message => $project . ' already in queue'}, status => IN_QUEUE)
+              if (!$job->{notes}{project_lock} || $job->{state} eq 'inactive');
+
+            $has_active_job = 1;
+        }
+    }
+    return $self->render(json => {message => 'queue full'}, status => QUEUE_FULL)
+      if ($results->{total} >= $queue_limit);
+
+    $app->gru->enqueue('obs_rsync_run', {project => $project}, {priority => 100});
```

Comment:
This looks wrong and i'm not sure why it would work, `enqueue` arguments should be an arrayref, not a hashref.

Reply:
Gru#enqueue expects hashref , then forms and passes arrayref to minion#enqueue

Reply:
because we can! :)